README.org
#+TITLE: SolarSailNavigator (Alpha)
#+OPTIONS: toc:nil
#+OPTIONS: num:nil
#+OPTIONS: author:nil

This is a developmental release of a solar sail and low thrust engine
navigation plugin for Kerbal Space Program (KSP,
http://www.kerbalspaceprogram.com). Plot the course of a spacecraft
pushed by the pressure of sunlight or a continuous thrust engine
(e.g. ion engines). Unlike rockets, solar sails and continuous thrust
engines apply a small amount of thrust throughout the trajectory,
gradually changing the orbit until reaching their destination. This
plugin simulates these propulsion systems and helps plan transfer
trajectories.

A sub-module of this project is called PersistentThrust and is
licensed under the KSP Interstellar open source license. It includes:
- Solar sail model with animated deployment by Sasquatch Marmalade
- Force model by Fractal_UK from the KSP Interstellar mod that works
  during time warp
- Force model of continuous thrust engines that during time warp
- Module Manager patch to the stock ion engine that reduces the thrust
  to realistic levels while keeping power consumption the same

The main plugin includes the Navigator module, which adds controls to
any command pod to schedule the steering and throttle of the vessel,
show what the future trajectory will look like, and show how close the
vessel gets to a target at the end of the command schedule. The
navigation system (everything outside PersistentThrust) is licensed
under the LGPL (see LICENSE.txt)

For the source code, go to: https://github.com/bld/SolarSailNavigator

This plugin adds critical features needed to navigate a solar sail or
continuous thrust engine spacecraft:

1. Set the vessel attitude to a reference frame that rotates with with
   the vessel as it orbits the sun. Specify two angles to point the
   sail, throttle, and duration. Add as many angle/angle/throttle/time
   maneuvers as you like to make a maneuver schedule. Time warp ahead,
   and the plugin will set the vessel's attitude and throttle.
   *WARNING:* Not recommended for planet centered orbits yet. The
   current steering strategy is really geared towards the center of
   gravity and sunlight being the same.
2. Preview the future trajectory and what the final orbit looks like
   in map mode, so you can try to plot a trajectory to a planet or
   other destination. Select a target and see how far off course you
   will be at the end of the command sequence. Warp ahead, and watch
   the sail follow your planned trajectory. The preview updates
   automatically as you change the command schedule.
3. Use multiple solar sails and persistent thrust engines. Use the
   "Use Sail" toggle to retract or deploy a sail for a segment. Set
   the throttle to 0 to disable any engines.
** Usage
*** Tutorial
See: https://github.com/bld/SolarSailNavigator/wiki/Tutorials
*** Building a solar sail spacecraft
In the Vehicle Assembly Building (VAB), look in the "Engines" section
for the "Solar Sail" part. The default orientation of the solar sail
part has the back (night) side of the sail pointing upwards, and the
front (sunlit) side facing down, so plan accordingly. When the sail is
deployed, it can shade the solar panels. You alse need to be able to
point the spacecraft in a variety of different directions to control
where it thrusts. So, put survival solar panels that can get light in
all directions - front, back, sides, up, down. Though, as we'll see
later, this plugin doesn't need power to change attitude and get power
if you run out. But, that's cheating.

You can deploy the sail on the launch pad to see how it will
look. DON'T lock the attidue on the pad, unless you like explosions.
*** Part GUI
Right clicking on a solar sail or persistent thrust engine will show
thrust, acceleration, and resource consumption (engine only). The sail
has a "Deploy Sail" button that will make the sail extend or retract,
and enable or disable it.
*** Command pod GUI
A "Show Navigator Controls" button will display the navigation tool
window.
*** Controls
**** Lock Attitude
There is a "Lock Attitude" toggle button. Clicking this will cause the
attitude maneuver profile to be executed, forcing the vessel to the
specified attitude, and setting the throttle (for the ion
engine). This is needed to hold the attitude and throttle during time
warp.
**** Start Time
This shows the time that the attitude maneuvers begin in Universal
Time (UT) seconds. Click the "Set to Now" button to reset the start
time to now. Do this if you are starting a new sequence.
**** Reference frame
The sail attitude is controlled relative to a reference frame that
rotates with the vessel as it travels around the sun. It is defined by
the vectors:
- X: from the sun to the sail (radial, "R")
- Y: perpendicular to R, lying in the orbit, in the direction of the
  orbit rotation (tangential, "T")
- Z: perpendicular to the orbit plane (normal, "N")
**** Cone
The "Cone" angle is the angle between the sun and the line
perpendicular to the surface of the sail (sail "normal" vector). As
the sail tilts away from the sun, the reflected light will start
generating thrust perpendicular to the sun-sail line. This is useful
for thrusting with the orbit velocity to sail away from the sun, or
against the velocity to tack closer to the sun. At 0 degrees, the sail
faces directly at the sun. At 90 degrees, the sail is edge on to the
sun, which effectively turns the sail off. This is the rotation about
the Z axis, after the clock angle has been applied.
**** Clock
The "Clock" angle is the rotation around the sun-sail line (X or R in
the RTN frame). It lets you direct the solar force above or below the
orbit plane to change the orbit inclination. If the clock angle is 0
degrees, the cone angle tilts the sail within the orbit plane. At
+/-90 degrees, the cone angle tilts the sail out of the orbit plane.
**** Flatspin
The "Flatspin" angle is the rotation about the spacecraft's local
up/down axis. Use this to, for example, make sure the solar panels are
oriented to collect power during maneuvers.
**** Throttle
The throttle (for continuous thrust engines) will set the vessel
throttle to the specified level from 0.00 to 1.00 in 0.05 increments
for the duration of the command segment.
**** Use Sail
The "Use Sail" toggle button indicates whether to deploy or retract
all sails on the vessel for this segment.
**** Days & Hours
This is the time to perform the attitude maneuver. Click +/- to
add/subtract one day, and +10/-10 to add/subtract 10 days. Click +/-
next to "Hours" fine tune the duration.
**** Color
This shows the color of this attitude maneuver segment on the
trajectory preview in map mode.
**** Add/Remove
The "Add" and "Remove" buttons add or subtract attitude maneuvers to
the sequence. The first maneuver cannot be removed.
**** Final Orbit Color
This shows the color of the final orbit, which is drawn in the preview.
**** Total
Total seconds of the entire attitude maneuver sequence.
*** Preview
The "Show Preview" button makes the plugin calculate and render the
future trajectory in map mode. The trajectory will be recalculated and
drawn whenever you change the angles, throttle, durations, or number
of maneuvers. This lets you incrementally change parameters and see
the effect on the trajectory. Once you've built an attitude maneuver
sequence that puts your final orbit where you want to go, increase the
time warp and watch the sail follow your trajectory. It will turn the
sail to 90 degrees and stop sailing once completed.
*** Final Orbit
Select "Show Final Orbit" to draw the final orbit on the map. Select
"Show Final Elements" to see the orbit elements of the final orbit.
*** Target
Select an object as a target in the map, and a line will be drawn
between the end of your trajectory and the closest approach to the
target. Select "Show Target Error" to see, in numbers, the distance
and velocity to the target, and the difference in your final orbit
elements with the target.
** Required Plugins
*** Module Manager
Module Manager is required to patch the ion engine part to use all the
features of this plugin, and adjust the thrust to realistic levels.
*** PersistentThrust
PersistentThrust is now its own separate module, which the navigator
uses to calculate the preview trajectory.
** Helpful Plugins
*** MechJeb
The Mechanical Jeb plugin (http://mechjeb.com) includes a tool called
"Warp Helper" that helps in executing the attitude maneuver
sequence. In the "Warp to" dropdown, select "Time". In the "Warp for"
field, type in the total duration, with "s" at the end for
seconds. Leave "Lead time" at 0s. Click "Warp" and watch it follow the
trajectory, returning to realtime and 90 degrees at the end.
** Known Problems and Limitations
See ISSUES.org and TODO.org
** Wish List
Also see TODO.org for items tagged WISHLIST
*** More sail models
One refinement that would make the current sail model more realistic
is to add more triangles to make it curved and billow away from the
sun.

Real sail designs like:
- IKAROS, the first solar sail, currently in orbit between Venus and
  Earth
- LightSail, the Planetary Society's cubesat solar sail, with launches
  in May, 2015 and in 2016
Sail concepts like
- Sunjammer (cancelled NASA solar sail test flight)
- Giant interplanetary sails like the Halley Rendezvous designs
- Ultrasail heliogyro
- Large, ultralight Interstellar Probe sail
- Giant interstellar laser-driven sails
*** Automated navigation
Manually building maneuver sequences can get you pretty close to your
destination. But, because of the near infinite variability of how to
steer a sail along the path to your destination and slow accumulation
of sail thrust, it can be difficult to reach it with any accuracy -
like rendezvous and orbital capture. For the moment, you may want a
secondary propulsion system to close the gap.

Professional solar sailors use numerical optimization algorithms to
calculate sail attitude vs. time for them. Manual sail planning can be
a good start for some algorithms, like gradient methods, which will
bring the sail in tight and accurate.
*** Automatic planetary & sun-centered sailing
There are simpler ways to automate sailing than the automated
navigation. There are steering strategies, for example, to
maximize/minimize the change in orbit elements for doing things like:
- Raising / escape
- Lowering / capture
- Inclination changes
- Precess an orbit
- Circularize / increase eccentricity
*** Realism
Real solar sails have a number of complex limitations that affect how
they sail.
**** Deployment
Real sails are too fragile to stow again after
deployment. Realistically, deployment could be a one time staged
event.
**** Pointing away from the sun
Some sail designs are naturally stable and have trouble pointing all
the way to 90 degrees. Limitations on how far away from the sun a sail
can point make life interesting for sail navigators.
**** Attitude control
Real sails use the torque of sunlight to steer. It would be
interesting exercise to have the sail slow down the time warp and use
special sail steering hardware to change attitude - slowly.
- Steering vanes
- Move the center of mass
- Thin film reflective control devices (used by IKAROS)

Real sails have massive moments of inertia, which make traditional
control by reaction wheels and RCS thrusters difficult.
**** Spinning sails
IKAROS spun to hold the sail flat instead of using structural booms.
**** Imperfect reflection
Real sails don't reflect light like a perfect, flat mirror. They are
curved, with wrinkles, and sail film that absorbs and scatters
light. The direction of thrust on a sail is affected by these
factors. Also, a non-flat sail will still present some area to the sun
at 90 degrees and produce a little thrust.
*** TweakScale
Get this plugin working with the sail to fly larger & smaller
sails by scaling the surface area (square scale change).


CHANGELOG.org
* SolarSailNavigator Changelog
** v1.0.8
*** Updated to compile with KSP 1.2
** v1.0.7
*** Updated to compile with KSP 1.1 and Unity 5
*** Navigator physics (attitude) updated in FixedUpdate() instead of OnFixedUpdate() to bypass part activation
*** Default "Sail on" and throttle now work
** v1.0.6
*** Don't use parts with PersistentEngine module but no ModuleEngines
*** Treat PersistentEngine as a PartModule with an engine field instead of treating it as a child of ModuleEngines
** v1.0.5
*** Popup window to change the default frame/angles/throttle/days/hours
*** PersistentThrust now a separate module: https://github.com/bld/PersistentThrust
*** Two reference frames now available: Radial/Tangential/Normal (RTN) and In-track/Cross-track/Normal (ICN)
**** RTN frame angles are Cone (angle off sun), Clock (rotation about sun), and Flatspin (rotation about vessel "up" vector)
**** ICN frame angles are Flight path angle, Azimuth (rotation about velocity), and Flatspin
*** Type in control angles, throttle, hours, and days
*** Fixed bug so that angles now limited to +/- 180 degrees
*** Fixed bug so that throttle now limited to 0-100%
*** Internally, uses new deltaV and mass consumption methods from PersistentThrust
** v1.0.4
*** Navigation controls now attached to a part on the spacecraft, instead of the propulsion system, and controls all continuous thrust parts on the vessel.
*** Realtime, timewarp, and navigation preview tested to work for multiple sails and persistent thrust engines.
*** "Use Sail" control added to retract or deploy the sail for specified command segments.
*** 3rd control angle "Flatspin" added to rotate spacecraft about its thrust axis to keep parts (e.g. solar panels) aligned as needed.
** v1.0.3
*** A new plugin called PersistentThrust is now included in the package. It implements the force models of the solar sail and continuous thrust engines.
*** The navigation plugin now supports both solar sails and continuous thrust engines.
*** Toggle buttons allow showing the preview, final orbit, final orbit elements, and difference between the final orbit and the target body.
*** Still unsupported:
**** Planetary trajectories
**** Other reference frames for steering
**** Multiple sails/engines
**** Propagation of the trajectory while not focused on the vessel
** v1.0.2
*** The line widths are now scaled to the distance from the camera in map view when drawing the sail trajectory preview, final orbit, and line to target object.
*** Now, you can specify the hours in an attitude control maneuver, in addition to days. This lets you have finer control over the trajectory.
*** Warp level has been removed as a control sequence option. It is now set to 100,000. Maybe later this will automatically adjust to short duration control sequences.
*** The plugin now includes a feature to help rendezvous with other objects. If a target object is selected:
**** Draw a line between the sail and target at the final time of the maneuver sequence
**** At the final time, show the difference between the sail and target of the:
***** Distance at final time
***** Speed at final time
***** Orbital elements
** v1.0.1
*** Small bug fix of v1.0.0-alpha. The trajectory propagation now samples the propagated orbits into the list to draw as a preview, instead of grabbing all. The criteria to sample one is to check if 1/360th of the current orbital period has passed since the last orbit was sampled.
*** Also, some of the preview-specific code has been pulled out of the sail module and put in the preview. This is in preparation for making a separate module to control & calculate the preview vs. the low thrust force model.
** v1.0.0
Initial Release


src/SailControls.cs
using System;
using System.Linq;
using UnityEngine;
using PersistentThrust;

namespace SolarSailNavigator {

    public class SailControl {

	// Fields
	// Cone Angle
	public float cone;
	public string cone_str;
	// Clock angle
	public float clock;
	public string clock_str;
	// Duration of steering maneuver
	public double duration;
	public string duration_str;
	// Time
	double days;
	string days_str;
	double hours;
	string hours_str;
	// Warp factor
	public int iwarp; // Index of warpLevels array
	public double warp; // Warp factor
	// Color
	public Color color;
	// Controls object this control is attached to
	public SailControls controls;
	// Sail this controls
	public SolarSailControlled sail;

	// Static fields
	public static float defaultCone = 90f;
	public static float defaultClock = 0f;
	public static double SecondsPerDay = 21600.0;
	public static double SecondsPerHour = 3600.0;
	public static double SecondsPerMinute = 60.0;
	public static double HoursPerDay = 6.0;
	public static double defaultDuration = 10 * SecondsPerDay;
	public static double[] warpLevels = { 1, 2, 3, 4, 5, 10, 50, 100, 1000, 10000, 100000 };
	public static int defaultiwarp = 10;
	
	// Cone controls
	public void GUICone () {
	    GUILayout.Label(cone_str, GUILayout.Width(30));
	    if (GUILayout.Button("+")) {
		cone += 5;
		if (cone > 90) {
		    cone = cone - 180;
		}
		cone_str = cone.ToString();
		controls.Update();
	    }
	    if (GUILayout.Button("-")) {
		cone -= 5;
		if (cone < -90) {
		    cone = cone + 180;
		}
		cone_str = cone.ToString();
		controls.Update();
	    }
	}

	// Clock controls
	public void GUIClock () {
	    GUILayout.Label(clock_str, GUILayout.Width(30));
	    if (GUILayout.Button("+")) {
		clock += 5;
		if (clock > 180) {
		    clock = clock - 360;
		}
		clock_str = clock.ToString();
		controls.Update();
	    }
	    if (GUILayout.Button("-")) {
		clock -= 5;
		if (clock < -180) {
		    clock = 360 + clock;
		}
		clock_str = clock.ToString();
		controls.Update();
	    }
	}

	// Time controls
	public void GUITime () {

	    // Days
	    GUILayout.Label(days_str, GUILayout.Width(30));
	    // Increase
	    GUILayout.BeginVertical();
	    if (GUILayout.Button("+")) {
		days++;
		days_str = days.ToString();
		duration = SecondsPerDay * days + SecondsPerHour * hours;
		duration_str = duration.ToString();
		controls.Update();
	    }
	    if (GUILayout.Button("+10")) {
		days += 10;
		days_str = days.ToString();
		duration = SecondsPerDay * days + SecondsPerHour * hours;
		duration_str = duration.ToString();
		controls.Update();
	    }
	    GUILayout.EndVertical();
	    // Decrease
	    GUILayout.BeginVertical();
	    if (GUILayout.Button("-")) {
		if (days > 0) {
		    days--;
		    days_str = days.ToString();
		    duration = SecondsPerDay * days + SecondsPerHour * hours;
		    duration_str = duration.ToString();
		    controls.Update();
		}
	    }
	    if (GUILayout.Button("-10")) {
		if (days >= 10) {
		    days -= 10;
		    days_str = days.ToString();
		    duration = SecondsPerDay * days + SecondsPerHour * hours;
		    duration_str = duration.ToString();
		    controls.Update();
		}
	    }
	    GUILayout.EndVertical();

	    // Hours
	    GUILayout.Label(hours_str, GUILayout.Width(10));
	    // Increase
	    if (GUILayout.Button("+")) {
		hours++;
		if (hours > HoursPerDay) {
		    hours = HoursPerDay;
		}
		hours_str = hours.ToString();
		duration = SecondsPerDay * days + SecondsPerHour * hours;
		duration_str = duration.ToString();
		controls.Update();
	    }
	    // Decrease
	    if (GUILayout.Button("-")) {
		if (hours > 0) {
		    hours--;
		    hours_str = hours.ToString();
		    duration = SecondsPerDay * days + SecondsPerHour * hours;
		    duration_str = duration.ToString();
		    controls.Update();
		}
	    }
	}

	// Color controls
	public void GUIColor (Color color) {
	    this.color = color;
	    var cstyle = new GUIStyle();
	    var ctx = new Texture2D(1, 1);
	    cstyle.normal.background = ctx;
	    ctx.SetPixel(1,1,color);
	    ctx.Apply();
	    GUILayout.Label(" ", cstyle, GUILayout.Width(30));
	}
	
	// GUI line
	public void GUILine (Color color) {
	    GUILayout.BeginHorizontal();

	    GUICone();
	    GUIClock();
	    GUITime();
	    GUIColor(color);

	    GUILayout.EndHorizontal();
	}
	
	// Parse a string to a single
	public static float ParseSingle (string str) {
	    float tmp;
	    if (Single.TryParse(str, out tmp)) {
		return tmp;
	    } else {
		return 0.0f;
	    }
	}

	// Parse a string to a double
	public static double ParseDouble (string str) {
	    double tmp;
	    if (Double.TryParse(str, out tmp)) {
		return tmp;
	    } else {
		return 0.0;
	    }
	}

	// Parse a string to an int
	public static int ParseInt (string str) {
	    int tmp;
	    if (Int32.TryParse(str, out tmp)) {
		return tmp;
	    } else {
		return 0;
	    }
	}

	// Constructor

	public SailControl(SolarSailControlled sail, SailControls controls, float cone, float clock, double duration, int iwarp) {
	    // Sail
	    this.sail = sail;
	    this.cone = cone;
	    // Parent controls object
	    this.controls = controls;
	    // Angles
	    cone_str = cone.ToString();
	    this.clock = clock;
	    clock_str = clock.ToString();
	    // Time
	    this.duration = duration;
	    this.days = Math.Floor(duration / SecondsPerDay);
	    this.days_str = days.ToString();
	    this.hours = Math.Floor(duration % SecondsPerDay / SecondsPerHour);
	    this.hours_str = this.hours.ToString();
	    this.duration = days * SecondsPerDay + hours * SecondsPerHour;
	    duration_str = duration.ToString();
	    // Warp factor
	    this.iwarp = iwarp;
	    warp = warpLevels[iwarp];
	}

	public static SailControl Default (SolarSailControlled sail, SailControls controls) {
	    return new SailControl (sail, controls, defaultCone, defaultClock, defaultDuration, defaultiwarp);
	}
    }

    public class SailControls {

	// Fields

	public int ncontrols;
	public SailControl[] controls;
	public SolarSailControlled sail;
	public double UT0;
	public SailControl sailOff;
	double durationTotal;
	public Color colorFinal;
	public bool showPreview = false;
	public bool showFinal = true;
	public bool showFinalElements = false;
	public bool showTargetErr = false;
	bool updateTargetLine = false;
	public SailPreview preview;
	public string previewButtonText = "Show Preview";
	
	// Static fields
	static Color[] colorMap = { Color.yellow,
				    Color.red,
				    Color.green,
				    Color.magenta,
				    Color.grey,
				    Color.blue,
				    Color.white };

	// Delimiter between controls
	private char delimiter = ':';

	// Constructor

	// Give the sail to which this control is for
	public SailControls (SolarSailControlled sail) {

	    // Assign sail field
	    this.sail = sail;
	    Debug.Log(this.sail.ToString());

	    // Initial time
	    if (sail.UT0 == 0) {
		UT0 = Planetarium.GetUniversalTime();
	    } else {
		UT0 = sail.UT0;
	    }
	    Debug.Log(UT0.ToString());

	    // Off sail control
	    sailOff = SailControl.Default(sail, this);

	    // If the sail doesn't have saved controls, return default
	    if (String.IsNullOrEmpty(sail.cones) ||
		String.IsNullOrEmpty(sail.clocks) ||
		String.IsNullOrEmpty(sail.durations)) {
		ncontrols = 1;
		controls = new SailControl[ncontrols];
		controls[0] = SailControl.Default(sail, this);

	    } else { // Otherwise, parse saved controls

		// Split into arrays
		var coneStrings = sail.cones.Split(delimiter);
		var clockStrings = sail.clocks.Split(delimiter);
		var durationStrings = sail.durations.Split(delimiter);

		// Find number of controls
		ncontrols = Math.Min(coneStrings.Length, Math.Min(clockStrings.Length, durationStrings.Length));

		// Initialize controls array
		controls = new SailControl[ncontrols];

		// Populate controls
		for(var i = 0; i < ncontrols; i++) {
		    controls[i] = new SailControl(sail,
						  this,
						  SailControl.ParseSingle(coneStrings[i]),
						  SailControl.ParseSingle(clockStrings[i]),
						  SailControl.ParseDouble(durationStrings[i]),
						  SailControl.defaultiwarp);
		}
	    }

	    // Preview
	    preview = new SailPreview(sail);
	}

	// Convert length in meters to string with bigger units
	string LengthToString(double lengthm) {
	    if (Math.Abs(lengthm) < 1000) {
		return Math.Round(lengthm).ToString() + " m";
	    } else if (Math.Abs(lengthm) < 1000000000) {
		return Math.Round(lengthm / 1000, 3).ToString() + " km";
	    } else {
		return Math.Round(lengthm / 1000000000, 4).ToString() + " Gm";
	    }
	}

	// Convert speed in m/s to string with bigger units
	string SpeedToString(double speedms) {
	    if (Math.Abs(speedms) < 1000) {
		return Math.Round(speedms, 1).ToString() + " m/s";
	    } else {
		return Math.Round(speedms / 1000, 3).ToString() + " km/s";
	    }
	}
	
	// GUI
	public void SailControlsGUI(int WindowID) {
	    GUILayout.BeginVertical();

	    // Lock/Unlock attitude
	    sail.IsLocked = GUILayout.Toggle(sail.IsLocked, "Lock Attitude");

	    // Set the initial time of the sequence
	    GUILayout.BeginHorizontal();
	    GUILayout.Label("Start time: " + UT0.ToString());
	    if(GUILayout.Button("Set to Now")) {
		UT0 = Planetarium.GetUniversalTime();
		Update();
	    }
	    GUILayout.EndHorizontal();

	    // Controls
	    GUILayout.BeginHorizontal();
	    GUILayout.Label("Cone", GUILayout.Width(80));
	    GUILayout.Label("Clock", GUILayout.Width(80));
	    GUILayout.Label("Days", GUILayout.Width(120));
	    GUILayout.Label("Hours", GUILayout.Width(65));
	    GUILayout.Label("Color", GUILayout.Width(30));
	    GUILayout.EndHorizontal();

	    int icolor = 0;
	    durationTotal = 0.0;
	    foreach(var control in controls) {

		// Draw individual GUIs
		control.GUILine(colorMap[icolor]);
		
		// Update total duration
		durationTotal += control.duration;

		// Update color index
		icolor++;
		if (icolor == colorMap.Length) {
		    icolor = 0;
		}
	    }
	    
	    // Add/remove control segments
	    GUILayout.BeginHorizontal();
	    if (GUILayout.Button("Add")) { Add(); };
	    if (GUILayout.Button("Remove")) { Remove(); };
	    GUILayout.EndHorizontal();

	    // Final orbit color
	    colorFinal = colorMap[icolor];
	    var cstyle = new GUIStyle();
	    var ctx = new Texture2D(1, 1);
	    cstyle.normal.background = ctx;
	    ctx.SetPixel(1,1,colorFinal);
	    ctx.Apply();
	    GUILayout.BeginHorizontal();
	    GUILayout.Label("Final orbit color: ");
	    GUILayout.Label(" ", cstyle);
	    GUILayout.EndHorizontal();

	    // Total duration of sequences
	    GUILayout.Label("Duration: " + durationTotal + " sec");

	    // Preview orbit
	    if (GUILayout.Toggle(showPreview, previewButtonText) != showPreview) {
		showPreview = !showPreview;
	    	if (showPreview) {
		    preview.Calculate();
		} else {
		    preview.Destroy();
		}
	    }

	    // If preview turned on
	    if (showPreview) {
		// Show final orbit
		if (GUILayout.Toggle(showFinal, "Show Final Orbit") != showFinal) {
		    showFinal = !showFinal;
		    preview.linef.enabled = showFinal;
		    Debug.Log("linef.enabled: " + preview.linef.enabled);
		}
		// Final elements
		GUILayout.BeginHorizontal();
		if (GUILayout.Toggle(showFinalElements, "Show Final Elements") != showFinalElements) {
		    showFinalElements = !showFinalElements;
		}
		if (showFinalElements) {
		    GUILayout.BeginVertical();
		    GUILayout.Label("ApA: " + LengthToString(preview.orbitf.ApA));
		    GUILayout.Label("PeA: " + LengthToString(preview.orbitf.PeA));
		    GUILayout.Label("TP: " + Math.Round(preview.orbitf.period, 2) + " sec");
		    GUILayout.Label("Inc: " + Math.Round(preview.orbitf.inclination, 2) + " deg");
		    GUILayout.Label("Ecc: " + Math.Round(preview.orbitf.eccentricity, 2));
		    GUILayout.Label("LAN: " + Math.Round(preview.orbitf.LAN, 2) + " deg");
		    GUILayout.Label("AOP: " + Math.Round(preview.orbitf.argumentOfPeriapsis, 2) + " deg");
		    GUILayout.EndVertical();
		}
		GUILayout.EndHorizontal();
	    }

	    // Target error
	    // Target line will need updating when target isn't selected
	    if (FlightGlobals.fetch.VesselTarget == null) {
		updateTargetLine = true;
	    }
	    if (FlightGlobals.fetch.VesselTarget != null && showPreview && preview.orbitf != null) {
		// Calculate target line & errors if target selected but line is null
		if (FlightGlobals.fetch.VesselTarget != null && updateTargetLine) {
		    preview.CalculateTargetLine();
		    updateTargetLine = false;
		}
		// GUI to show target errors
		GUILayout.BeginHorizontal();
		if (GUILayout.Toggle(showTargetErr, "Show Target Error") != showTargetErr) {
		    showTargetErr = !showTargetErr;
		}
		if (showTargetErr) {
		    GUILayout.BeginVertical();
		    // Distance
		    GUILayout.Label("Final distance to target: " + LengthToString(preview.targetD));
		    // Speed
		    GUILayout.Label("Final speed to target: " + SpeedToString(preview.targetV));
		    // Orbit elements
		    GUILayout.Label("Apoapsis: " + LengthToString(preview.ApErr));
		    GUILayout.Label("Periapsis: " + LengthToString(preview.PeErr));
		    GUILayout.Label("Orbital Period: " + Math.Round(preview.TPErr, 2) + " sec");
		    GUILayout.Label("Inclination: " + Math.Round(preview.IncErr, 3) + " deg");
		    GUILayout.Label("Eccentricity: " + Math.Round(preview.EccErr, 3));
		    GUILayout.Label("LAN: " + Math.Round(preview.LANErr, 3));
		    GUILayout.Label("AOP: " + Math.Round(preview.AOPErr, 3));
		    GUILayout.EndVertical();
		}
		GUILayout.EndHorizontal();
	    }

	    /*
	    // Show difference with target at end of control sequence
	    if (FlightGlobals.fetch.VesselTarget != null && showPreview) {
		// Distance
		GUILayout.Label("Final distance to target: " + LengthToString(preview.targetD));
		
		// Speed
		GUILayout.Label("Final speed to target: " + SpeedToString(preview.targetV));
	    }
	    */
	    
	    GUILayout.EndVertical();

	    GUI.DragWindow();
	}

	// Controls GUI rectangle
	private Rect controlWindowPos = new Rect(0, 50, 0, 0);

	// Controls GUI function
	public void DrawControls () {
	    if (sail.vessel == FlightGlobals.ActiveVessel)
		controlWindowPos = GUILayout.Window(10, controlWindowPos, SailControlsGUI, "Sail Controls");
	}
	
	// Add a control
	public void Add () {
	    var newControls = new SailControl[ncontrols + 1];
	    for(var i = 0; i < ncontrols; i++) {
		newControls[i] = controls[i];
	    }
	    newControls[ncontrols] = SailControl.Default(sail, this);
	    controls = newControls;
	    ncontrols++;
	    colorFinal = colorMap[ncontrols % colorMap.Length];
	    Update();
	}
	
	// Remove a control
	public void Remove () {
	    if (ncontrols > 1) { // Don't remove last control
		var newControls = new SailControl[ncontrols - 1];
		for (var i = 0; i < ncontrols - 1; i++) {
		    newControls[i] = controls[i];
		}
		controls = newControls;
		ncontrols--;
		colorFinal = colorMap[ncontrols % colorMap.Length];
		Update();
	    }
	}
	
	// Return control at specified UT
	public SailControl Lookup (double UT) {
	    double deltaUT = UT - UT0; // Time past UT0
	    double durationSum = 0.0; // Running sum of durations
	    foreach (var control in controls) {
		// deltaUT between prior and next duration
		if (deltaUT >= durationSum && deltaUT < (durationSum += control.duration)) {
		    return control;
		}
	    }
	    // If none found, return "sail off" control
	    return sailOff;
	}

	// Update the sail's persistant sail control fields
	public void Update () {
	    // Initial time
	    sail.UT0 = UT0;
	    // Controls
	    sail.cones = controls[0].cone_str;
	    sail.clocks = controls[0].clock_str;
	    sail.durations = controls[0].duration_str;
	    for (var i = 1; i < ncontrols; i++) {
		sail.cones += delimiter + controls[i].cone_str;
		sail.clocks += delimiter + controls[i].clock_str;
		sail.durations += delimiter + controls[i].duration_str;
	    }
	    Debug.Log(sail.UT0.ToString());
	    Debug.Log(sail.cones);
	    Debug.Log(sail.clocks);
	    Debug.Log(sail.durations);

	    preview.Calculate();
	}
    }
}
